lib/services/workload_identity_x509_revocation.go

```diff
@@ -0,0 +1,133 @@
+	// WorkloadIdentityX509Revocation.
+	UpsertWorkloadIdentityX509Revocation(
+		ctx context.Context,
+		workloadIdentityX509Revocation *workloadidentityv1pb.WorkloadIdentityX509Revocation,
+	) (*workloadidentityv1pb.WorkloadIdentityX509Revocation, error)
+}
+
+// MarshalWorkloadIdentityX509Revocation marshals the
+// WorkloadIdentityX509Revocation object into a JSON byte array.
+func MarshalWorkloadIdentityX509Revocation(
+	object *workloadidentityv1pb.WorkloadIdentityX509Revocation, opts ...MarshalOption,
+) ([]byte, error) {
+	return MarshalProtoResource(object, opts...)
+}
+
+// UnmarshalWorkloadIdentityX509Revocation unmarshals the
+// WorkloadIdentityX509Revocation object from a JSON byte array.
+func UnmarshalWorkloadIdentityX509Revocation(
+	data []byte, opts ...MarshalOption,
+) (*workloadidentityv1pb.WorkloadIdentityX509Revocation, error) {
+	return UnmarshalProtoResource[*workloadidentityv1pb.WorkloadIdentityX509Revocation](data, opts...)
+}
+
+// ValidateWorkloadIdentityX509Revocation validates the
+// WorkloadIdentityX509Revocation object.
+// It returns a nil if the object is valid, otherwise an error.
+func ValidateWorkloadIdentityX509Revocation(s *workloadidentityv1pb.WorkloadIdentityX509Revocation) error {
+	switch {
+	case s == nil:
+		return trace.BadParameter("object cannot be nil")
+	case s.Version != types.V1:
+		return trace.BadParameter("version: only %q is supported", types.V1)
+	case s.Kind != types.KindWorkloadIdentityX509Revocation:
+		return trace.BadParameter("kind: must be %q", types.KindWorkloadIdentityX509Revocation)
+	case s.Metadata == nil:
+		return trace.BadParameter("metadata: is required")
+	case s.Metadata.Name == "":
+		return trace.BadParameter("metadata.name: is required")
+	case s.Metadata.Expires == nil:
+		return trace.BadParameter("metadata.expires: is required")
+	case s.Metadata.Expires.IsValid() == false:
+		return trace.BadParameter("metadata.expires: must be valid")
+	case s.Metadata.Expires.AsTime().IsZero():
+		return trace.BadParameter("metadata.expires: must be non-zero")
+	case s.Spec == nil:
+		return trace.BadParameter("spec: is required")
+	case s.Spec.Reason == "":
+		return trace.BadParameter("spec.reason: is required")
+	case s.Spec.RevokedAt == nil:
+		return trace.BadParameter("spec.revoked_at: is required")
+	case s.Spec.RevokedAt.IsValid() == false:
+		return trace.BadParameter("spec.revoked_at: must be valid")
+	case s.Spec.RevokedAt.AsTime().IsZero():
+		return trace.BadParameter("spec.revoked_at: must be non-zero")
+	}
+
+	// Name must be a integer encoded as hex - this is the serial number of the
+	// X509 cert. Whilst typically presented using a colon separated hex string,
+	// here we will remove the colons. We will also ensure it is encoded in
+	// lowercase, to ensure consistency.
+	serial := big.Int{}
+	_, ok := serial.SetString(s.Metadata.Name, 16)
+	if !ok {
+		return trace.BadParameter("metadata.name: must be a hex encoded integer without colons")
+	}
```

Comment:
Match against `(?:[0-9a-f]{2})`? This also doesn't check that the length is even, which we might want to do I think?

If you're not feeling the regexp, instead of loading the number into a `big.Int` why not use `encoding/hex.DecodeString`?

Reply:
cd0f45a0da